### DISCUSSION TITLE
Is StatefulSession  thread safe ?

### DISCUSSION THREAD

**Discussion by kuldeeparora89** [evrete/evrete#89] [General]
Hi,

I have a scenario where multiple threads needs to execute insertAndFire on same rule set over different data . Is session class thread safe or we need to create separate session for each thread ?

**Comment by andbi** [comment]
The default implementation is not thread-safe. Calling `insertAndFire` on a stateful session from different threads may produce unpredictable results. It is also worth noting that, generally speaking, the state of stateful sessions may depend on the order of the `insertAndFire` calls.

If your rules produce the expected results when given a collection of facts, then creating another session for another dataset is the way to go.

Thanks for bringing up the thread safety topic; I'll update the docs.

**Comment by kuldeeparora89** [comment]
Thanks @andbi  , we implemented session per thread model for our use case.
